### PR TITLE
[CAP 35] Update XDR defintions and xdr package for CAP35

### PR DESCRIPTION
**Pull request by 2opremio**
Using the patch from https://github.com/stellar/stellar-protocol/blob/master/core/cap-0035.md#xdr-changes

Part of #3348

### CODE REVIEW

File: xdr/asset.go [hunk 1]
```diff
@@ -45,8 +45,8 @@ func MustNewCreditAsset(code string, issuer string) Asset {
 }
 
 // NewAllowTrustAsset returns a new allow trust asset, panicking if it can't.
-func NewAllowTrustAsset(code string) (AllowTrustOpAsset, error) {
-	a := AllowTrustOpAsset{}
+func NewAllowTrustAsset(code string) (AssetCode, error) {
```
Comment (leighmcculloch) [inline xdr/asset.go:48]: I think this should be renamed to `NewAssetCode` to continue to match the name of the type it is creating? We can keep the old function too for backwards compatibility if that is useful.

File: xdr/asset.go [hunk 2]
```diff
@@ -67,7 +67,7 @@ func NewAllowTrustAsset(code string) (AllowTrustOpAsset, error) {
 }
 
 // MustNewAllowTrustAsset returns a new allow trust asset, panicking if it can't.
-func MustNewAllowTrustAsset(code string) AllowTrustOpAsset {
+func MustNewAllowTrustAsset(code string) AssetCode {
```
Comment (leighmcculloch) [inline xdr/asset.go:70]: I think this should be renamed to `MustNewAssetCode` to continue to match the name of the type it is creating?We can keep the old function too for backwards compatibility if that is useful.

File: xdr/asset.go [hunk 3]
```diff
@@ -222,22 +222,22 @@ func (a *Asset) SetNative() error {
 
 // ToAllowTrustOpAsset for Asset converts the Asset to a corresponding XDR
 // "allow trust" asset, used by the XDR allow trust operation.
-func (a *Asset) ToAllowTrustOpAsset(code string) (AllowTrustOpAsset, error) {
+func (a *Asset) ToAllowTrustOpAsset(code string) (AssetCode, error) {
 	length := len(code)
```
Comment (leighmcculloch) [inline xdr/asset.go:226]: I think this should be renamed to `ToAssetCode` to continue to match the name of the type it is creating?We can keep the old function too for backwards compatibility if that is useful.

File: xdr/xdr_generated.go [hunk 4]
```diff
@@ -8656,159 +8994,20 @@ var (
...
-//        case ASSET_TYPE_CREDIT_ALPHANUM4:
-//            AssetCode4 assetCode4;
-//
-//        case ASSET_TYPE_CREDIT_ALPHANUM12:
-//            AssetCode12 assetCode12;
-//
-//            // add other asset types here in the future
-//        }
-//
-type AllowTrustOpAsset struct {
```
Comment (leighmcculloch) [inline xdr/xdr_generated.go:8673]: We can use Go type aliases to define that AllowTrustOpAsset is the same as AssetCode for backwards compatibility, if that is useful. If that is not useful then I wouldn't worry about it.

**Review by leighmcculloch — APPROVED** [review]
🎉 Looks great.

I left some suggestions (💡) as comments inline, but I defer to you for all of them to you for whether they're the right call or not as I'm not sure if backwards compatibility is important for this change, and not sure what other plans you might have.

FYI the XDR is likely to undergo at least one more change that isn't yet captured in the diff in the CAP-35 document. That change is discussed on the mailing list as item 4 in this message: https://groups.google.com/g/stellar-dev/c/hPhkXhrl5-Y/m/Hi9XaWhkCwAJ.

File: xdr/asset.go (hunk 1 above)
Comment (2opremio) [reply xdr/asset.go:48]: Makes sense, will do

File: xdr/asset.go (hunk 3 above)
Comment (2opremio) [reply xdr/asset.go:226]: good point

File: xdr/xdr_generated.go (hunk 4 above)
Comment (2opremio) [reply xdr/xdr_generated.go:8673]: Also a good point. I am not sure it's worth maintaining backwards compatibility here. @stellar/horizon-committers thoughts?

Comment (leighmcculloch) [reply xdr/xdr_generated.go:8673]: I'm not convinced it's worth it either. I think I've seen us do major version releases for new protocols before?

File: xdr/asset.go (hunk 1 above)
Comment (2opremio) [reply xdr/asset.go:48]: I had to use `NewAssetCodeFromString` because xdrgen already generated `NewAssetCode`

File: xdr/asset.go (hunk 2 above)
Comment (2opremio) [reply xdr/asset.go:70]: I used `MustNewAssetCodeFromString` for consistency with the change above

File: xdr/asset.go (hunk 3 above)
Comment (2opremio) [reply xdr/asset.go:226]: Done

File: xdr/xdr_generated.go (hunk 4 above)
Comment (2opremio) [reply xdr/xdr_generated.go:8673]: I think I will merge now. @stellar/horizon-committers please let me know if you think I should change this in terms of backwards compatibility.

I will also add a changelog entry indicating the changes.

**Pull request merged**